### PR TITLE
feat: update gene annotation script

### PR DESCRIPTION
**Pull request by pecholleyc**
### Main improvements in this PR:
Improve the python script used to generate the ID mapping table (`ensembl_ID_mapping.tsv`):

- the script writes the `genes.tsv` file directly.
- the column names have changed following the decisions in issue #157
- the two new columns of PR  #244 are generated and included in the output file
- The script can now be imported as a module and be used in other python script
- The newly generated `gene.tsv` fixes issue #250

The idea is also to generate the `genes.tsv` file when the sanityCheck.py complains on the genes IDs comparison. A PR that updates the `sanityCheck.py` will be created, once this PR has been accepted.

**I hereby confirm that I have:**

- [ ] Tested my code on my own computer for running the model
- [X] Selected `develop` as a target branch

### CODE REVIEW

File: code/GPRs/fetch_ensembl_gene_annotations.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+import requests
+import re
+from collections import OrderedDict
+
+"""
+Fetch and write in OUTPUT FILE gene annotations of the input YAML file. The file must contains Ensembl IDs as gene IDs.
+Input:
+ - the YAML file format of the GEM
+ - version of Ensembl, e.g. 103
+ - version of Ensembl, e.g. 103
```
Comment (mihai-sysbio) [inline code/GPRs/fetch_ensembl_gene_annotations.py:14]: This line appears duplicated
```suggestion
```

**Comment by mihai-sysbio** [comment]
Really nice stuff @pecholleyc! Random thought: would it be too much work to generalize the script for other (non-human) models?

**Comment by pecholleyc** [comment]
There is maybe some additional works in Matlab script because of the removal of the `ensembl_ID_mapping.tsv` :neutral_face:

**Comment by JonathanRob** [comment]
> There is maybe some additional works in Matlab script because of the removal of the `ensembl_ID_mapping.tsv` 😐

Yep, I'm working on these changes and will update them soon.

**Comment by pecholleyc** [comment]
The script is already 'kind of' generalized for non-human model, there is a `--database` argument that can be used to target any model in Ensembl. But I never tested with another DB, I hope the main SQL queries is still valid. The problem will be maybe with the query that fetches the primary assembly IDs; It can be tricky to generalize because of the name of the chromosomes are different between species.

**Comment by Hao-Chalmers** [comment]
> There is maybe some additional works in Matlab scripts because of the removal of the ensembl_ID_mapping.tsv 😐

a random thought: since a new script "fetch_ensembl_gene_annotations.py" was introduced here, the removal of `get_ensembl_ID_mapping.py` and `ensembl_ID_mapping.tsv`, as well as the derived adjustments, are not necessarily a part of this branch.

**Comment by pecholleyc** [comment]
@Hao-Chalmers, The new script supplant the old one and the previous logic, I don't think both should exist at the same time in the repository. I think it will make things more complicated.

**Comment by JonathanRob** [comment]
@Hao-Chalmers Also, I've already started the work on this branch, it would be easier to just continue it here.

**Comment by Hao-Chalmers** [comment]
> @Hao-Chalmers, The new script supplant the old one and the previous logic, I don't think both should exist at the same time in the repository. I think it will make things more complicated.

@pecholleyc true, they should not both exist in the `master` branch. While the `develop` suppose to be dynamic and with tolerance.

> @Hao-Chalmers, The new script supplant the old one and the previous logic, I don't think both should exist at the same time in the repository. I think it will make things more complicated.

@JonathanRob then go ahead

**Comment by nanjiangshu** [comment]
In the YAML file `Human-GEM.yml`, `version` is empty and the term `organism` is missing. Is that done by purpose or?

**Comment by mihai-sysbio** [comment]
> In the YAML file `Human-GEM.yml`, `version` is empty and the term `organism` is missing. Is that done by purpose or?

The `version` is blank on `develop` and filled in on the main branch.

**Comment by nanjiangshu** [comment]
OK, for the `organism`, it can be inferred from the newly added term `taxonomy: "9606"`

**Comment by Hao-Chalmers** [comment]
@nanjiangshu which branch are you talking about?

**Comment by JonathanRob** [comment]
Ok, I made a few changes:

1. fd94216: Updated the `translateGrRules` function so that it uses the `genes.tsv` annotation file instead of the now-deprecated `ensembl_ID_mapping.tsv` file 

2. 7c2d4f0: Updated the `annotateGEM` function to be compatible with the new `genes.tsv` file, and to retrieve more of the fields (columns) in the other annotation tsv files.

3. 983382d: I noticed that the Rhea IDs that were recently added to the `reactions.tsv` file had the format `RHEA:#####`. This is inconsistent with how we typically store the IDs in the annotation TSVs (i.e., without the prefix). I trimmed the `REAH:` prefixes from the identifiers in the `reactions.tsv` file.

**Review by JonathanRob — APPROVED** [review]

**Comment by mihai-sysbio** [comment]
> is is inconsistent with how we typically store the IDs in the annotation TSVs (i.e., without the prefix).

Is there any exception besides the ChEBI prefix in `metabolites.tsv` (which has to stay there)?

**Comment by nanjiangshu** [comment]
> @nanjiangshu which branch are you talking about?

Hi @Hao-Chalmers , thanks for your reply. I mean the branch of this pull request, i.e., `feat/update_gene_annotation_script`.
@mihai-sysbio has alreadly answered my question in the post above. https://github.com/SysBioChalmers/Human-GEM/pull/251#issuecomment-823956404

**Comment by JonathanRob** [comment]
> Is there any exception besides the ChEBI prefix in `metabolites.tsv` (which has to stay there)?

@mihai-sysbio as far as I know, ChEBI is the only weird case that doesn't follow the same url pattern as the other identifiers.

**Comment by Hao-Chalmers** [comment]
looks good to me

**Comment by pecholleyc** [comment]
I don't fully approve the removal of the prefix of Rhea IDs because, similar to ChEBI IDs, it is part of the identifier. I obtain them by parsing Rhea and retrieved the values provided by the `accession` key. This format is, I believe, used by other databases providing mapping to Rhea (e.g. UniProt). The problem is that Rhea is just using ##### as `id` internally (in their database), and also in their URLs. Maybe this is probably the reason it has been registered on Identifiers.org to accept only digits as ID (which is inaccurate in my opinion).

**Comment by JonathanRob** [comment]
@pecholleyc Ok, I see. It seems the ID formatting is more complicated than I thought, and is not consistent among different sources.

In that case, we should undo my removal of the "RHEA:" prefix, and I can incorporate the ID adjustment into the `annotateGEM` function, as I did with the ChEBI ID.

**Comment by mihai-sysbio** [comment]
@pecholleyc I think you're right, see the screenshot below when trying to match a KEGG reaction id on the Rhea-db website:

image

Even Rhea-db presents its identifiers with the `RHEA:` prefix. It's very much in line with what ChEBI does, so @JonathanRob's suggestion makes sense to me.

**Pull request merged**